R/plot.r

```diff
@@ -36,42 +45,55 @@
 #'   evaluation.
 #' @export
 #' @examples
-#' # Generate some sample data, then compute mean and standard deviation
-#' # in each group
+#' # Create a data frame with some sample data, then create a data frame
+#' # containing the mean value for each group in the sample data.
 #' set.seed(1)
-#' df <- data.frame(
-#'   gp = factor(rep(letters[1:3], each = 10)),
-#'   y = rnorm(30)
+#' 
+#' sample_df <- data.frame(
+#'   group = factor(rep(letters[1:3], each = 10)),
+#'   value = rnorm(30)
 #' )
-#' ds <- do.call(rbind, lapply(split(df, df$gp), function(d) {
-#'   data.frame(mean = mean(d$y), sd = sd(d$y), gp = d$gp)
-#' }))
-#'
-#' # The summary data frame ds is used to plot larger red points on top
-#' # of the raw data. Note that we don't need to supply `data` or `mapping`
-#' # in each layer because the defaults from ggplot() are used.
-#' ggplot(df, aes(gp, y)) +
+#' 
+#' group_means_df <- setNames(
+#'   aggregate(value ~ group, sample_df, mean),
+#'   c("group", "group_mean")
+#' )
+#' 
+#' # The following three code blocks create the same graphic, each using one
+#' # of the three patterns specified above. In each graphic, the group means
+#' # data frame is used to plot larger red points on top of the sample data.
+#' 
+#' # Pattern 1
+#' # Both the `data` and the `mapping` arguments are passed into the `ggplot()`
+#' # call. Note that we don't need to supply `mapping` or `data` arguments in
+#' # the first `geom_point()` layer because those arguments get passed along
+#' # from the `ggplot()` call.
+#' ggplot(data = sample_df, mapping = aes(x = group, y = value)) +
 #'   geom_point() +
-#'   geom_point(data = ds, aes(y = mean), colour = 'red', size = 3)
-#'
-#' # Same plot as above, declaring only the data frame in ggplot().
-#' # Note how the x and y aesthetics must now be declared in
-#' # each geom_point() layer.
-#' ggplot(df) +
-#'   geom_point(aes(gp, y)) +
-#'   geom_point(data = ds, aes(gp, mean), colour = 'red', size = 3)
-#'
-#' # Alternatively we can fully specify the plot in each layer. This
-#' # is not useful here, but can be more clear when working with complex
-#' # mult-dataset graphics
+#'   geom_point(mapping = aes(x = group, y = group_mean), data = group_means_df,
```

Comment:
I think it is a feature of the pattern to omit `x = group` here,  because that aesthetic is inherited from the `ggplot()` call.

Reply:
I see your point, but I was putting a lot of emphasis on the parallelism across the examples. 
Could we add a comment to the effect that the `x = group` could also be omitted because the it would be passed along from the `ggplot()` call?

```
# Pattern 1
# Both the `data` and the `mapping` arguments are passed into the `ggplot()`
# call. Note that we don't need to supply `mapping` or `data` arguments in
# the first `geom_point()` layer because those arguments get passed along
# from the `ggplot()` call. Note also that the `x = group` argument in the 
# second `geom_point()` could also be omitted because it would also be 
# passed along from the `ggplot() call.
ggplot(data = sample_df, mapping = aes(x = group, y = value)) +
  geom_point() +
  geom_point(mapping = aes(x = group, y = group_mean), data = group_means_df,
             colour = 'red', size = 3
  )
```

Reply:
I think it might be better to omit `x = group` here, and then note in the comment that we're allowed to do so because it has a global mapping.

Reply:
How about this:

    # Pattern 1
    # Both the `data` and the `mapping` arguments are passed into the `ggplot()`
    # call. Note that we don't need to supply `mapping` or `data` arguments in
    # the first `geom_point()` layer because those arguments get passed along
    # from the `ggplot()` call.
    ggplot(data = sample_df, mapping = aes(x = group, y = value)) +
      geom_point()
    
    # Pattern 2
    # Same plot as above, passing in only the `data` argument in the `ggplot()`
    # call. Note how we must now pass the `mapping` argument in for the
    # `geom_point()` layer because it is no longer passed along from the 
    # `ggplot()` call.
    ggplot(data = sample_df) +
      geom_point(mapping = aes(x = group, y = value))
    
    # Pattern 3
    # Same plot as above, but with larger red points on top of the sample data
    # to show the mean values. Here, we are not passing either the `data` or
    # `mapping` arguments into the `ggplot()` call. We must then fully specify 
    # those arguments in each layer. This pattern can be particularly useful
    # when creating more complex, multi-data frame graphics.
    ggplot() +
      geom_point(mapping = aes(x = group, y = value), data = sample_df) +
      geom_point(mapping = aes(x = group, y = group_mean), data = group_means_df,
                 colour = 'red', size = 3
      )

This removes the second layer from the first two patterns, which maybe helps focus reader on the differences in the patterns. The third pattern then adds in the additional layer with the red points for the mean values, which is also in the direction of showing the pattern being used for a "more complex, multi-data frame" graphic.  

The outstanding issue, then (I think) is whether the second layer in Pattern 3 should have the code wrap or have a line length pushing the 80 char limit but show better alignment across the arguments.

Reply:
I can see how that simplifies things, but maybe a bit too much as the body text in the details sections talks about using pattern 1 for adding a layer using data from another data frame. I think it is important to show this. Let me suggest the following:

```r
# Pattern 1
# Both the `data` and the `mapping` arguments are passed into the `ggplot()`
# call. The`mapping` or `data` arguments are omitted in the first `geom_point()` 
# layer because those arguments get passed along from the `ggplot()` call. 
# Note that the the second point layer re-uses the `x = group` mapping through
# that mechanism, but overrides the y-aesthetic.
ggplot(data = sample_df, mapping = aes(x = group, y = value)) +
  geom_point() +
  geom_point(
    mapping = aes(y = group_mean), data = group_means_df,
    colour = 'red', size = 3
  )

# Pattern 2
# Same plot as above, passing in only the `data` argument in the `ggplot()`
# call. The `mapping` arguments are now required for each
# `geom_point()` layer because there is no global `mapping` argument passed
# along from the `ggplot()` call.
ggplot(data = sample_df) +
  geom_point(mapping = aes(x = group, y = value)) +
  geom_point(
    mapping = aes(x = group, y = group_mean), data = group_means_df,
    colour = 'red', size = 3
  )

# Pattern 3
# Same plot as above, passing in neither the `data` or `mapping` arguments
# to the `ggplot()` call. Both those arguments must then be fully specified
# in each layer. While not useful here, this pattern can be useful when 
# creating complex graphics with multiple datasets.
ggplot() +
  geom_point(mapping = aes(x = group, y = value), data = sample_df) +
  geom_point(
    mapping = aes(x = group, y = group_mean), data = group_means_df,
    colour = 'red', size = 3
  )
```

Reply:
Thanks @thomasp85 and @teunbrand! 

My most recent commit wraps those long lines, and has some edits in the comment blocks for each of the three patterns to attempt to use consistent phrasing across the examples.